src/irmin-pack/unix/chunked_suffix.ml

```diff
@@ -0,0 +1,189 @@
+      Ao.open_rw ~path ~end_poff ~dead_header_size ~auto_flush_threshold
+        ~auto_flush_procedure
+
+    let open_ro ~root ~chunk_idx ~end_poff ~dead_header_size ~is_appendable =
+      let open Result_syntax in
+      let path = path ~root ~chunk_idx in
+      (* Passing [end_poff] is required by Ao for its consistency check (and
+         only applies to the appendable chunk).
+         TODO: We could change its API in the future so we can not do
+         [Io.size_of_path], which is admittedly very silly.*)
+      let* end_poff =
+        if is_appendable then Ok end_poff else Io.size_of_path path
+      in
+      Ao.open_ro ~path ~end_poff ~dead_header_size
+  end
+
+  let create_rw ~root ~start_idx ~overwrite ~auto_flush_threshold
+      ~auto_flush_procedure =
+    let open Result_syntax in
+    let chunk_idx = start_idx in
+    let+ ao =
+      Ao_chunk.create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure
+    in
+    let chunk = { idx = chunk_idx; suffix_off = Int63.zero; ao } in
+    let inventory = Inventory.v 1 (Fun.const chunk) in
+    { inventory }
+
+  let open_rw ~root ~end_poff ~start_idx ~chunk_num ~dead_header_size
+      ~auto_flush_threshold ~auto_flush_procedure =
+    let open Result_syntax in
+    let open_chunk ~chunk_idx ~is_appendable =
+      match is_appendable with
+      | true ->
+          Ao_chunk.open_rw ~root ~chunk_idx ~end_poff ~auto_flush_threshold
+            ~auto_flush_procedure ~dead_header_size
+      | false ->
+          Ao_chunk.open_ro ~root ~chunk_idx ~end_poff ~dead_header_size
+            ~is_appendable
+    in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let open_ro ~root ~end_poff ~dead_header_size ~start_idx ~chunk_num =
+    let open Result_syntax in
+    let open_chunk = Ao_chunk.open_ro ~root ~end_poff ~dead_header_size in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let appendable_ao t = (Inventory.appendable t.inventory).ao
+  let end_poff t = appendable_ao t |> Ao.end_poff
+
+  let read_exn t ~off ~len buf =
+    let chunk = Inventory.find ~off t.inventory in
+    let poff = suffix_off_to_chunk_poff chunk off in
+    Ao.read_exn chunk.ao ~off:poff ~len buf
+
+  let append_exn t s = Ao.append_exn (appendable_ao t) s
+
+  let close t =
+    (* Close immutable chunks, ignoring errors. *)
+    let _ =
+      t.inventory
+      |> Inventory.iter_skip_appendable @@ fun chunk ->
+         let _ = Ao.close chunk.ao in
+         ()
+    in
+    (* Close appendable chunk and keep error since this
+       is the one that can have a pending flush. *)
+    appendable_ao t |> Ao.close
+
+  let empty_buffer t = appendable_ao t |> Ao.empty_buffer
+  let flush t = appendable_ao t |> Ao.flush
+  let fsync t = appendable_ao t |> Ao.fsync
+
+  let refresh_end_poff t new_end_poff =
+    Ao.refresh_end_poff (appendable_ao t) new_end_poff
```

Comment:
My understanding is that this the end poff for the last chunk and not the length of the full suffix. The control file also tracks the end poff for the last chunk. 
We only need the length of the full suffix in the dispatcher, for which an additional function is needed. I suggested to add it in the my PR https://github.com/mirage/irmin/pull/2115#discussion_r997138686 but we can add it here if you prefer (I don't think it matters much, since there is no split and no tests here).

Reply:
Yeah, we should keep track of these somehow, either TODOs or as a comment with checkboxes in the main issue. I'll do something to help us remember.

Adding to what @icristescu says: for `end_poff`, as mentioned in https://github.com/mirage/irmin/pull/2115#discussion_r995922877, I do think we ought to rename it in chunked suffix and in control file (`consistency_poff`, `persisted_poff`, ???) to try to make its purpose clearer. I think a pair of functions, `Chunked_suffix.consistency_poff` and `Chunked_suffix.refresh_consistency_poff`, would make it clearer, even if they are just delegating to the last chunk's `end_poff`.

As for the length of the suffix, I am okay to have it in the split PR (with a suggestion for a rename :smile: https://github.com/mirage/irmin/pull/2118#discussion_r998604496).